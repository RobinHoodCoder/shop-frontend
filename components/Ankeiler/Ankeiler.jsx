import React from 'react';
import ItemStyles from '../styles/ItemStyles';
import PriceTag from '../styles/PriceTag';
import formatMoney from '../../lib/formatMoney';
import LinkBlock from '../styles/LinkBlock';
import TitleBlock from '../TitleBlock/TitleBlock';
import DeleteProduct from '../DeleteProduct/DeleteProduct';
import EditProduct from '../EditProduct/EditProduct';
const Ankeiler = (props) => {
  const { name, description, photo, price, id } = props;

  const { altText = '', image } = photo || {};
  const imageSrc = image?.publicUrlTransformed || '';

  return (
    <ItemStyles>
      <LinkBlock
        href={`/product/${id}`}
      >
        <img
          alt={altText}
          src={imageSrc}
        />
        {price && (
          <PriceTag>
            {formatMoney(price)}
          </PriceTag>
        )}
        <TitleBlock
          name={name}
          description={description}
        />
      </LinkBlock>
      <div className="buttonList">
        <DeleteProduct
          id={id}
          name={name}
        >
          🗑️ Delete
        </DeleteProduct>
        <EditProduct id={id}>
          🖊️ Edit
        </EditProduct>
      </div>
    </ItemStyles>
  );
};

export default Ankeiler;
