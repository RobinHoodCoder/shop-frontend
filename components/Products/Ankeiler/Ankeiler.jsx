import React from 'react';
import ItemStyles from '../../styles/ItemStyles';
import PriceTag from '../../styles/PriceTag';
import formatMoney from '../../../lib/formatMoney';
import LinkBlock from '../../styles/LinkBlock';
import TitleBlock from '../../TitleBlock/TitleBlock';
import DeleteProduct from '../../DeleteProduct/DeleteProduct';
import EditProduct from '../../EditProduct/EditProduct';
import { AddToCartButton } from '../../CartButtons/AddToCartButton/AddToCartButton';
const Ankeiler = (props) => {
  const { name, description, photo, price, id } = props;

  const { altText = '', image } = photo || {};
  const imageSrc = image?.publicUrlTransformed || '';

  return (
    <ItemStyles>
      <LinkBlock
        href={`/product/${id}`}
      >
        <img
          alt={altText}
          src={imageSrc}
        />
        {price && (
          <PriceTag>
            {formatMoney(price)}
          </PriceTag>
        )}
        <TitleBlock
          name={name}
          description={description}
        />
      </LinkBlock>
      <div className="buttonList">
        <DeleteProduct
          id={id}
          name={name}
        >
          🗑️
        </DeleteProduct>
        <EditProduct id={id}>
          🖊️
        </EditProduct>
        <AddToCartButton id={id}/>
      </div>
    </ItemStyles>
  );
};

export default Ankeiler;
