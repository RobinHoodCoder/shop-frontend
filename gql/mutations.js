import gql from 'graphql-tag';

export const M_CREATE_PRODUCT = gql`
    mutation CREATE_PRODUCT_MUTATION (
        $name: String!
        $description: String!
        $price: Int!
        $image: Upload
    ) {
        createProduct(
            data: {
                name: $name
                description: $description
                price: $price
                status: "AVAILABLE"
                photo: {
                    create: {
                        image: $image,
                        altText: $name
                    }
                }
            }
        ) {
            id
            price
            description
            name
            id
        }
    }
`;

export const M_UPDATE_PRODUCT = gql`
  mutation UPDATE_PRODUCT (
    $id: ID!
    $name: String!
    $description: String!
    $price: Int!
#    $image: Upload
  ) {
    updateProduct(
      id: $id
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
#        photo: {
#          create: {
#            image: $image,
#            altText: $name
#          }
#        }
      }
    ) {
      id
      price
      description
      name
      id
    }
  }
`;

export const M_DELETE_PRODUCT = gql`
  mutation DELETE_PRODUCT (
    $id: ID!
  ) {
    deleteProduct(
      id: $id
    ) {
      name
      id
    }
  }
`;

export const M_LOGIN = gql`
    mutation LOGIN (
        $email: String!
        $password: String!
    ) {
        authenticateUserWithPassword(
            email: $email
            password: $password
        ) {
           ...on UserAuthenticationWithPasswordSuccess {
             item {
                 email
                 name
             }
           }
           ...on UserAuthenticationWithPasswordFailure {
            code
            message
           }
        }
    }
`;

export const M_LOGOUT = gql`
  mutation LOGOUT {
    endSession
  }
`;


export const M_SIGNUP = gql`
  mutation SIGNUP_MUTATION (
    $email: String!
    $password: String!
    $name: String!
  ) {
    createUser(
    data: {
      name: $name
      email: $email
      password: $password
    }
    ) {
        email
        id
        name
    }
  }
`;

export const M_REQUEST_RESET_PASSWORD = gql`
  mutation REQUEST_RESET_PASSWORD_MUTATION (
    $email: String!
  ) {
    sendUserPasswordResetLink(
      email: $email
    ) {
      code
      message
    }
  }
`;

export const M_RESET_PASSWORD = gql`
    mutation RESET_PASSWORD_MUTATION (
        $email: String!
        $token: String!
        $password: String!
    ) {
        redeemUserPasswordResetToken(
            email: $email
            token: $token
            password: $password
        ) {
            code
            message
        }
    }
`;

export const M_ADD_TO_CART = gql`
    mutation ADD_TO_CART_MUTATION($id: ID!) {
        addToCart(productId: $id) {
            id
        }
    }
`;

export const M_REMOVE_FROM_CART = gql`
    mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
        deleteCartItem(id: $id) {
            id
        }
    }
`;

export const M_CREATE_ORDER = gql`
    mutation M_CREATE_ORDER($token: String!) {
        checkout(token: $token) {
            id
            charge
            total
            items {
                id
                name
            }
        }
    }
`;
