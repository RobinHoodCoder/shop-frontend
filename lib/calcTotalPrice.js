import formatMoney from './formatMoney';

const calcTotalPrice = (cart) => {
  return cart.reduce((currentOutput, cartItem) => {
    console.log(currentOutput, cartItem);
    const { product } = cartItem || {};
    if (!product) { // check if it's available because product can be deleted
      return currentOutput;
    }

    if (!product.price) {
      return currentOutput;
    }
    return currentOutput + (cartItem.quantity * product.price);
  }, 0);
};

export const formattedTotalPrice = (cart) => {
  return formatMoney(
    calcTotalPrice(cart)
  );
};

export default calcTotalPrice;
