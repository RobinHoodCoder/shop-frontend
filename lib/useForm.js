import { useEffect, useState } from 'react';

export const useForm = (initial = {}) => {
  const [formValues, setInputs] = useState(initial);
  const handleChange = (e) => {
    const { value, name, type, files } = e.target;

    let outVal = '';

    switch (type) {
    case 'number' :
      outVal = !!value ? parseInt(value, 10) : 0;
      break;
    case 'file' :
      // console.log(value);
      [outVal] = files;
      break;
    default:
      outVal = value;
    }

    setInputs((prevState) => {
      return {
        ...prevState,
        [name]: outVal,
      };
    });
  };
  const resetForm = (e) => {
    e.preventDefault();
    setInputs(initial);
  };
  const clearForm = (e) => {
    e.preventDefault();
    const blankArr = Object.entries(initial).map(([key, _]) => {
      return [key, ''];
    });
    const blankObj = Object.fromEntries(blankArr);
    setInputs(blankObj);
  };

  useEffect(() => {

  }, []);

  return { formValues, handleChange, resetForm, clearForm };
};
