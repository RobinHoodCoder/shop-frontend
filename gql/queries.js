import gql from 'graphql-tag';

export const Q_ALL_PRODUCTS = gql`
    query ALL_PRODUCTS ( $first: Int = 1, $skip: Int = 0, ) 
    {
    allProducts (
     first: $first
     skip: $skip
    ) {
        id
        name
        description
        price
        photo {
            altText
            id
            image
            {
                publicUrlTransformed
            }
        }
    }
}
`;
export const Q_CURRENT_USER = gql`
    query  {
    authenticatedItem {
        ...on User {
                id
                email
                name
                # Query the curruser once we have it
            }
        }
    }
`;

export const Q_SINGLE_PRODUCT = gql`
    query SINGLE_PRODUCT (
       $id: ID!
    ) 
    {
        Product(where: {id: $id}) {
            id
            name
            description
            price
            photo {
                altText
                id
                image
                {
                    publicUrlTransformed
                }
            }
        }
    }
`;
export const Q_PAGINATION = gql`
    query PAGINATION_QUERY
    {
        _allProductsMeta {
            count
        }
    }
`;
