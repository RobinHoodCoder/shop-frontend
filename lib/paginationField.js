import gql from 'graphql-tag';

const paginationField = () => {
  console.log('paginationField');
  return {
    keyArgs: false,
    read(
      existing = [], // e.g. }]
      {
        args, // gimme the first x things  and skip 2
        cache, // reference to the cache from Apollo
      } // e.g. :
    ) {
      const { skip, first } = args;
      // Read number of items from cache

      const data = cache.readQuery({
        Q_PAGINATION: {
          skip,
          first,
        },
      });
      const count = data?._allProductsMeta?.count;

      const page = (skip / first) + 1;
      const pages = Math.ceil(count / first);
      // ask read func for the items

      // look for non-existing items
      const items = existing.slice(skip, skip + 1);

      if (items.length !== first) {
        // We dont have all the items, so we need to fetch them
        // console.log(existing, args, cache, data);
        return false;
      }
      // If there are items, first return from cache, and we don' t need to go to the network.
      if (items.length) {
        console.log(`There are ${items.length} items in the cache. we gonna send em to Apollo.`);
        return items;
      }
    },
    merge(newThings) {
      console.log(newThings);
      // Runs when Apollo client turns back from the network.
    },
  };
};

export default paginationField;
