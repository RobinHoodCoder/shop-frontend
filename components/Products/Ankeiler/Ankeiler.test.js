
import { render, screen } from '@testing-library/react';
import Ankeiler from './Ankeiler';
import { MockedProvider } from '@apollo/client/testing';
import { fakeItem } from '../../../lib/testUtils';
import { Q_SINGLE_PRODUCT } from '../../../gql/queries';

const product = fakeItem();

const mocks = [
  {
    // When someone requests this query and variable combo
    request: {
      query: Q_SINGLE_PRODUCT,
      variables: {
        id: '123',
      },
    },
    // Return this data
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe(`<Product/> price`, () => {
  it(`Renders the component`, () => {
    expect(true)
      .toBe(true);

    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <Ankeiler {...product}>Test</Ankeiler>
      </MockedProvider>,
    );
    expect(screen.getByTestId('price', { exact: false })).toBeInTheDocument();
    debug();
  });
});

describe(`<Product/> markup matches snapshot`, () => {
  it(`Renders the component, matches the snapshot`, () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <Ankeiler {...product}>Test</Ankeiler>
      </MockedProvider>,
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByTestId('price', { exact: false }))
      .toBeInTheDocument();
    debug();
  });
});
