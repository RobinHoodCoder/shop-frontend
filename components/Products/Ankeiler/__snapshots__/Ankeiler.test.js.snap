// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> markup matches snapshot Renders the component, matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles__Item-sc-16pk14u-0 fAxeyb"
  >
    <div
      class="LinkBlock__LinkBlockStyles-sc-1vgdqsk-0 gKDMie"
    >
      <div>
        <img
          alt="dogs are best"
          src="dog.jpg"
        />
        <span
          class="PriceTag-sc-nwbk6t-0 iCGkBe"
          data-testid="price"
        >
          €50
        </span>
        <div
          class="TitleBlock__TitleBlockStyles-sc-1oa93hk-0 EzZnr"
        >
          <h3
            class="Title-sc-16nq74k-0 jKqyec"
          >
            <span>
              dogs are best
            </span>
          </h3>
          <p
            class="Description-sc-1etfdho-0 jCgNpW"
          >
            <span>
              dogs
            </span>
          </p>
        </div>
      </div>
    </div>
    <div
      class="buttonList"
    >
      <button>
        🗑️
      </button>
      <div>
        <a
          href="/edit?id=abc123"
        >
          🖊️
        </a>
      </div>
      <button
        type="button"
      >
        Add
         To Cart 🛒
      </button>
    </div>
  </div>
</div>
`;

exports[`<Product/> matches snapshot Renders the component, matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles__Item-sc-16pk14u-0 fAxeyb"
  >
    <div
      class="LinkBlock__LinkBlockStyles-sc-1vgdqsk-0 gKDMie"
    >
      <div>
        <img
          alt="dogs are best"
          src="dog.jpg"
        />
        <span
          class="PriceTag-sc-nwbk6t-0 iCGkBe"
          data-testid="price"
        >
          €50
        </span>
        <div
          class="TitleBlock__TitleBlockStyles-sc-1oa93hk-0 EzZnr"
        >
          <h3
            class="Title-sc-16nq74k-0 jKqyec"
          >
            <span>
              dogs are best
            </span>
          </h3>
          <p
            class="Description-sc-1etfdho-0 jCgNpW"
          >
            <span>
              dogs
            </span>
          </p>
        </div>
      </div>
    </div>
    <div
      class="buttonList"
    >
      <button>
        🗑️
      </button>
      <div>
        <a
          href="/edit?id=abc123"
        >
          🖊️
        </a>
      </div>
      <button
        type="button"
      >
        Add
         To Cart 🛒
      </button>
    </div>
  </div>
</div>
`;
